File: packages/js/src/RulesetMerge.ts
import type { FieldRule, PatternRule, ValidatorSpec, WriteLynxRuleset } from "@writelynx/builtin-rules";

export namespace RulesetMerge {
    /**
     * Merges a partial ruleset overlay onto a base ruleset (custom overrides builtin by id).
     *
     * @param base - Base ruleset (typically builtin)
     * @param overlay - Partial custom ruleset
     * @returns New merged ruleset
     */
    export function merge(base: WriteLynxRuleset, overlay: Partial<WriteLynxRuleset>): WriteLynxRuleset {
        const validators: Record<string, ValidatorSpec> = { ...base.validators };

        if (overlay.validators) {
            Object.assign(validators, overlay.validators);
        }

        const fieldRules = mergeRulesById(base.field.rules, overlay.field?.rules ?? []);
        const patternRules = mergeRulesById(base.pattern.rules, overlay.pattern?.rules ?? []);

        return {
            validators,
            field: { rules: fieldRules },
            pattern: { rules: patternRules }
        };
    }

    /**
     * Merges rule arrays by id; overlay entries replace base entries with the same id.
     *
     * @param baseRules - Base rule list
     * @param overlayRules - Overlay rule list
     * @returns Merged rule list
     */
    function mergeRulesById<T extends { id: string }>(baseRules: T[], overlayRules: T[]): T[] {
        const byId = new Map<string, T>();

        for (const rule of baseRules) {
            byId.set(rule.id, rule);
        }

        for (const rule of overlayRules) {
            byId.set(rule.id, rule);
        }

        return Array.from(byId.values());
    }

    /**
     * Removes a rule by id from field and pattern lists.
     *
     * @param ruleset - Active ruleset
     * @param id - Rule id to remove
     * @returns Updated ruleset without the rule
     */
    export function removeRuleById(ruleset: WriteLynxRuleset, id: string): WriteLynxRuleset {
        return {
            validators: { ...ruleset.validators },
            field: {
                rules: ruleset.field.rules.filter((r: FieldRule) => r.id !== id)
            },
            pattern: {
                rules: ruleset.pattern.rules.filter((r: PatternRule) => r.id !== id)
            }
        };
    }
}

File: packages/lua/src/writelynx/ruleset_merge.lua
local M = {}

--- Merges rule arrays by id.
--- @param base_rules table[]
--- @param overlay_rules table[]
--- @return table[]
local function merge_rules_by_id(base_rules, overlay_rules)
    local by_id = {}

    for _, rule in ipairs(base_rules) do
        by_id[rule.id] = rule
    end

    for _, rule in ipairs(overlay_rules) do
        by_id[rule.id] = rule
    end

    local result = {}

    for _, rule in pairs(by_id) do
        table.insert(result, rule)
    end

    return result
end

--- Merges overlay ruleset onto base.
--- @param base table
--- @param overlay table
--- @return table
function M.merge(base, overlay)
    local validators = {}

    for k, v in pairs(base.validators or {}) do
        validators[k] = v
    end

    if overlay.validators then
        for k, v in pairs(overlay.validators) do
            validators[k] = v
        end
    end

    return {
        validators = validators,
        field = {
            rules = merge_rules_by_id(base.field.rules, overlay.field and overlay.field.rules or {})
        },
        pattern = {
            rules = merge_rules_by_id(base.pattern.rules, overlay.pattern and overlay.pattern.rules or {})
        }
    }
end

return M
